skyjo_immutable.py

```diff
@@ -0,0 +1,513 @@
+
+    def copy(self) -> Self:
+        return Deck(_cards=list(self._cards), _rng=self._rng)
+
+
+@dataclasses.dataclass(slots=True, frozen=True)
+class DiscardPile:
+    discarded_card_counts: np.array = dataclasses.field(
+        default_factory=lambda: np.zeros(NUM_CARD_TYPES)
+    )
+    top_card: typing.Optional[int] = None
+
+    def copy(self) -> Self:
+        return DiscardPile(
+            discarded_card_counts=self.discarded_card_counts.copy(),
+            top_card=self.top_card,
+        )
+
+    def discard(self, card: int) -> Self:
+        next_discarded_card_counts = self.discarded_card_counts.copy()
+        next_discarded_card_counts[card + 2] += 1
+        return dataclasses.replace(
+            self, discarded_card_counts=next_discarded_card_counts, top_card=card
+        )
+
+    def replace_top_card(self, card: int) -> Self:
+        next_discarded_card_counts = self.discarded_card_counts.copy()
+        next_discarded_card_counts[self.top_card + 2] -= 1
+        next_discarded_card_counts[card + 2] += 1
+        return dataclasses.replace(
+            self, discarded_card_counts=next_discarded_card_counts, top_card=card
+        )
+
+
+@dataclasses.dataclass(slots=True, frozen=True)
+class ImmutableSkyjoState:
+    num_players: int = dataclasses.field()
+    """Number of players in the game"""
+    player_scores: np.array = dataclasses.field()
+    """Current player scores"""
+    hands: list[Hand] = dataclasses.field(default=None)
+    """The list of participating players."""
+    deck: Deck = dataclasses.field(default=None)
+    """Deck of cards"""
+    discard_pile: DiscardPile = dataclasses.field(default_factory=DiscardPile)
+    """Discard Pile Object"""
+    drawn_card: int | None = None
+    """Card drawn (if applicable)"""
+    curr_player: int = 0
+    """Index of current player"""
+    turn_count: int = 0
+    """Number of turns taken so far"""
+    round_count: int = 0
+    """Number of rounds so far"""
+    is_round_ending: bool = 0
+    """If this is the last turn for each player"""
+    round_ending_player: int | None = None
+    """Index of player who has ended the game"""
+    valid_actions: list[SkyjoAction] = dataclasses.field(default_factory=list)
+    """List of valid actions available from current state"""
+    winning_player: int | None = None
+    """Index of player who won the game"""
+
+    def setup_round(self) -> Self:
+        next_deck = Deck(_cards=DECK, _rng=self.deck._rng).shuffle()
+        next_hands, next_deck = next_deck.deal_hands(self.num_players)
+        top_card, next_deck = next_deck.draw_top_card()
+        next_hands = [hand.initial_flips() for hand in next_hands]
+        starting_player = self.round_ending_player
+        if self.round_ending_player is None:
+            hand_values = [hand.visible_points() for hand in next_hands]
+            starting_player = hand_values.index(max(hand_values))
+        next_valid_actions = next_hands[
```

Comment:
maybe needless to always compute next valid actions (constructing objects is expensive (though probably not relative to training))